File: SVN_REPO/midterm2/q07/numOfLetters/src/main/kotlin/numOfLetters.kt
fun countLetters(string: String): Map<Char, Int> {
    tailrec fun countLettersRec(
        remaining: String,
        counts: Map<Char, Int>
    ): Map<Char, Int> {
        return if (remaining.isEmpty()) {
            counts
        } else {
            val currentLetter = remaining[0]
            val updatedCounts = counts + (currentLetter to (counts.getOrDefault(currentLetter, 0) + 1))
            countLettersRec(remaining.substring(1), updatedCounts)
        }
    }

    return countLettersRec(string, emptyMap())
}

fun printLetterCounts(letterCounts: Map<Char, Int>) {
    letterCounts.forEach { (letter, count) ->
        println("$letter: $count times")
    }
}

fun main() {
    print("Enter the string: ")
    val userInput = readLine().orEmpty()

    val letterCounts = countLetters(userInput)
    printLetterCounts(letterCounts)
}

File: SVN_REPO/midterm2/q06/numOfLetters/src/numOfLetters.rb
def count_letters(string, counts = Hash.new(0))
  return counts if string.empty?

  current_letter_count = counts.merge(string[0] => counts[string[0]] + 1)
  count_letters(string[1..-1], current_letter_count)
end

def print_letter_counts(letter_counts)
  letter_counts.each do |letter, count|
    puts "#{letter}: #{count} times"
  end
end

print "Enter the string: "
user_input = gets.chomp

letter_counts = count_letters(user_input)
print_letter_counts(letter_counts)


